### DISCUSSION TITLE
Importing JavaScript from other file/npm package throws error

### DISCUSSION THREAD

**Discussion by TheGreatViolet** [caoccao/Javet#170] [Q&A] [answered]
Hey there! I've been trying to migrate my project from J2V8 to Javet for a while now. However a bit of a snag I've hit is that when I try to import JS code from another file/from a package such as axios, Javet throws an error along the lines of:

`com.caoccao.javet.exceptions.JavetExecutionException: Error: Cannot find package './importTestModule.js'`

I'm just wondering if this could be something I'm doing wrong or if Javet doesn't support this or if it could be something different entirely. I've read through the documentation and other issues/discussions to try to find an answer to no avail. I'd be happy to share more code if needed to. Thanks!

**Comment by caoccao** [comment, marked as answer]
May I know which mode you are using?

# V8 Mode

You need to register a module resolver via `v8Runtime.setV8ModuleResolver(...)`. Please refer to this [test suite](https://github.com/caoccao/Javet/blob/main/src/test/java/com/caoccao/javet/values/reference/TestV8Module.java) for detail.

# Node.js Mode

You need to follow some rules to make it work. Please refer to #148 for detail.

**Reply by TheGreatViolet** [reply]
I'm running Javet in Node.js mode, and executing my script from a string. The issue you lead me to I had already seen, and most of the things in it I had already done prior, however it caused me take another look at the documentation on modularization, which lead me to manually setting the require root directory by calling `.getNodeModule(NodeModuleModule.class).setRequireRootDirectory(getResourceFile("jsTests").getAbsolutePath())` on my NodeRuntime. However when I do so I get this error:

`Unexpected exception thrown: java.lang.NullPointerException: Cannot invoke "com.caoccao.javet.interop.IV8Native.createV8Value(long, int, Object)" because "this.v8Native" is null`

I'm not entirely sure what the issue is. I've looked at some obvious things, such as the directory not existing but everything seems fine. Here's the code I'm using to instantiate the runtime:

```java
try (JavetEnginePool<NodeRuntime> pool = new JavetEnginePool<NodeRuntime>()) {
            pool.getConfig().setJSRuntimeType(JSRuntimeType.Node);

            IJavetEngine<NodeRuntime> engine = pool.getEngine();
            NodeRuntime runtime1 = engine.getV8Runtime();

            // Set require root directory to the jsTests folder inside the resources folder

            runtime1.getNodeModule(NodeModuleModule.class).setRequireRootDirectory(getResourceFile("jsTests").getAbsolutePath());

            // Set class runtime to runtime1

            runtime = runtime1;
        } catch (JavetException e) {
            BlockJS.instance.getLogger().log(Level.SEVERE,
                    "There was an error creating the Javet runtime.");
            throw new RuntimeException(e);
        }
```

**Reply by caoccao** [reply]
Please specify the pool type via the constructor and try again.

**Reply by TheGreatViolet** [reply]
I'm assuming by this you meant to specify a JavetEngineConfig object in the constructor for the JavetEnginePool with the JSRuntimeType set to Node. This is what I came up with:

```java
JavetEngineConfig config = new JavetEngineConfig();

config.setJSRuntimeType(JSRuntimeType.Node);

try (JavetEnginePool<NodeRuntime> pool = new JavetEnginePool<NodeRuntime>(config)) {
    pool.getConfig().setJSRuntimeType(JSRuntimeType.Node);

    IJavetEngine<NodeRuntime> engine = pool.getEngine();
    NodeRuntime runtime1 = engine.getV8Runtime();

    // Set require root directory to the jsTests folder inside the resources folder

    runtime1.getNodeModule(NodeModuleModule.class).setRequireRootDirectory(getResourceFile("jsTests").getAbsolutePath());

    // Set class runtime to runtime1

    runtime = runtime1;
} catch (JavetException e) {
    BlockJS.instance.getLogger().log(Level.SEVERE,
            "There was an error creating the Javet runtime.");
    throw new RuntimeException(e);
}
```

Still with these changes made it still throws the same `Cannot invoke "com.caoccao.javet.interop.IV8Native.createV8Value(long, int, Object)" because "this.v8Native" is null` error as before.

**Reply by caoccao** [reply]
You cannot store the runtime for usage outside the try-with-resource because the runtime is closed. It's similar to InputStream / OutputStream that you cannot use them after they are closed.

**Reply by caoccao** [reply]
Also, are you trying that on Android? There is no Node.js mode for Android.

**Reply by TheGreatViolet** [reply]
Ok I fixed the runtime so now it doesn't throw the NullPointerException now, however now it throws the `Error: Cannot find package './importTestModule.js'` from before. Here's my code now:

```java
NodeRuntime setupRuntime() throws JavetException {
        JavetEngineConfig config = new JavetEngineConfig();

        config.setJSRuntimeType(JSRuntimeType.Node);

        JavetEnginePool<NodeRuntime> pool = new JavetEnginePool<NodeRuntime>(config);

        pool.getConfig().setJSRuntimeType(JSRuntimeType.Node);

        NodeRuntime runtime = pool.getEngine().getV8Runtime();

        runtime.getNodeModule(NodeModuleModule.class).setRequireRootDirectory(getResourceFile("jsTests").getAbsolutePath());

        return runtime;
    }
```

And here's the test I'm using the runtime in:

```java
@Test
    @DisplayName("Can run JS with imports")
    void runJSWithImports() {
        try (NodeRuntime runtime = setupRuntime()) {
            runner = new JSExecutionInterface(getResourceFile("jsTests/importTest.js"), runtime);
            runner.runtime.await();
            assertDoesNotThrow(() -> runner.registerAndRun());
        } catch (JavetException e) {
            BlockJS.instance.getLogger().log(Level.SEVERE,
                    "There was an error creating the Javet runtime.");
            throw new RuntimeException(e);
        }
    }
```

To answer your other question, no I'm not running this on Andrioid, I'm running it on Linux.

**Reply by caoccao** [reply]
# V8 Mode

You need to register a module resolver via `v8Runtime.setV8ModuleResolver(...)`. Please refer to this [test suite](https://github.com/caoccao/Javet/blob/main/src/test/java/com/caoccao/javet/values/reference/TestV8Module.java) for detail.

**Reply by TheGreatViolet** [reply]
Setting the module resolver totally worked! Here's the code I ended up with for the module resolver:

```java
@Test
    @DisplayName("Can run JS with imports")
    void runJSWithImports() {
        try (NodeRuntime runtime = setupRuntime()) {
            runtime.setV8ModuleResolver(((v8Runtime, resourceName, iv8Module) -> {
                if (resourceName.equals("./importTestModule.js")) {
                    return runtime.getExecutor(getResourceFile("jsTests/importTestModule.js"))
                            .setResourceName(resourceName)
                            .compileV8Module();
                } else {
                    return null;
                }
            }));

            runner = new JSExecutionInterface(getResourceFile("jsTests/importTest.js"), runtime);
            runner.runtime.await();
            assertDoesNotThrow(() -> runner.registerAndRun());
        } catch (JavetException e) {
            BlockJS.instance.getLogger().log(Level.SEVERE,
                    "There was an error creating the Javet runtime.");
            throw new RuntimeException(e);
        }
    }
```

Thanks so much for your help in resolving this issue, I've been stuck on it for weeks.

**Reply by caoccao** [reply]
Great to see that working. And thank you for sharing the solution with the community.

**Comment by zahrahariry** [comment]
I have the same problem. With the difference that I have to add a module not just single file . Thank you for your guidance.
i have to use v8 mode not Node.js mode.

**Reply by caoccao** [reply]
It's the same solution. Your module resolver needs to handle the requests respectively just like what Node.js does.
